File: examples/python/document_ttl.py
#!/usr/bin/env python3
"""
TTL Expiration Verification Test

This test VERIFIES that TTL actually works by:
1. Inserting a document with short TTL (3 seconds)
2. Verifying the document exists immediately
3. Waiting for TTL to expire
4. Verifying the document is GONE

If this test passes, TTL expiration is working correctly.
"""

import asyncio
import os
import sys
import time
from pathlib import Path
from dotenv import load_dotenv

from ekodb_client import Client

# Load environment variables
env_path = Path(__file__).parent.parent.parent / ".env"
load_dotenv(env_path)


async def main():
    print("╔════════════════════════════════════════════════════════╗")
    print("║     TTL EXPIRATION VERIFICATION TEST                   ║")
    print("╚════════════════════════════════════════════════════════╝")
    print()
    print("This test verifies that document TTL expiration works correctly.")
    print("We will insert documents with short TTL and verify they expire.")
    print()

    base_url = os.getenv("API_BASE_URL", "http://localhost:8080")
    api_key = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")

    client = Client.new(base_url, api_key)
    print("✓ Client connected")

    collection = "ttl_expiration_test_py"
    ttl_seconds = 3

    try:
        # ═══════════════════════════════════════════════════════════════════════
        # TEST 1: Document TTL Expiration
        # ═══════════════════════════════════════════════════════════════════════
        print()
        print("═══════════════════════════════════════════════════════════")
        print("TEST 1: Document TTL Expiration")
        print("═══════════════════════════════════════════════════════════")

        # Step 1: Insert document with 3-second TTL
        print(f"\n[Step 1] Insert document with {ttl_seconds} second TTL")
        print("  Input: {name: 'TTL Test', value: 'should expire'}")
        print(f"  TTL: {ttl_seconds}s")

        doc = await client.insert(
            collection,
            {"name": "TTL Test", "value": "should expire"},
            ttl=f"{ttl_seconds}s",
        )
        doc_id = doc.get("id")
        print(f"  Output: Document ID = {doc_id}")
        print("  ✓ PASS: Document inserted")

        # Step 2: Verify document exists immediately
        print("\n[Step 2] Verify document exists immediately")
        print(f"  Input: find_by_id({doc_id})")

        found = await client.find_by_id(collection, doc_id)
        if not found:
            raise Exception("❌ FAILED: Document should exist but was not found")
        name_val = found.get("name", {})
        if isinstance(name_val, dict):
            name_val = name_val.get("value", name_val)
        print(f"  Output: Found document with name = {name_val}")
        print("  ✓ PASS: Document exists")

        # Step 3: Wait for TTL to expire
        wait_time = ttl_seconds + 2
        print(f"\n[Step 3] Wait for TTL to expire ({wait_time}s)")
        print("  Waiting", end="", flush=True)
        for _ in range(wait_time):
            time.sleep(1)
            print(".", end="", flush=True)
        print(" done")
        print("  ✓ PASS: Wait complete")

        # Step 4: Verify document is GONE
        print("\n[Step 4] Verify document has expired")
        print(f"  Input: find_by_id({doc_id})")

        try:
            expired = await client.find_by_id(collection, doc_id)
            if expired is None:
                print("  Output: None (document not found)")
                print("  ✓ PASS: Document expired correctly!")
            else:
                print(f"  Output: Document still exists! {expired}")
                raise Exception(
                    "❌ FAILED: Document should have expired but still exists!"
                )
        except Exception as e:
            if "FAILED" in str(e):
                raise
            print(f"  Output: Error (expected) - {e}")
            print("  ✓ PASS: Document expired (not found error)")

        # ═══════════════════════════════════════════════════════════════════════
        # Cleanup
        # ═══════════════════════════════════════════════════════════════════════
        print("\n═══════════════════════════════════════════════════════════")
        print("CLEANUP")
        print("═══════════════════════════════════════════════════════════")

        await client.delete_collection(collection)
        print("✓ Deleted test collection")

        # ═══════════════════════════════════════════════════════════════════════
        # Summary
        # ═══════════════════════════════════════════════════════════════════════
        print()
        print("╔════════════════════════════════════════════════════════╗")
        print("║              ALL TTL TESTS PASSED ✓                    ║")
        print("╚════════════════════════════════════════════════════════╝")
        print()
        print("TTL expiration is working correctly:")
        print("  • Documents with TTL expire after the specified time")
        print("  • Documents without TTL persist indefinitely")
        print("  • Different TTL durations are handled correctly")

    except Exception as error:
        print(f"\n❌ TEST FAILED: {error}")
        try:
            await client.delete_collection(collection)
        except:
            pass
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: examples/go/document_ttl.go
// TTL Expiration Verification Test
//
// This test VERIFIES that TTL actually works by:
// 1. Inserting a document with short TTL (3 seconds)
// 2. Verifying the document exists immediately
// 3. Waiting for TTL to expire
// 4. Verifying the document is GONE
//
// If this test passes, TTL expiration is working correctly.

package main

import (
	"errors"
	"fmt"
	"log"
	"os"
	"time"

	ekodb "github.com/ekoDB/ekodb-client-go"
	"github.com/joho/godotenv"
)

func run() (runErr error) {
	fmt.Println("╔════════════════════════════════════════════════════════╗")
	fmt.Println("║     TTL EXPIRATION VERIFICATION TEST                   ║")
	fmt.Println("╚════════════════════════════════════════════════════════╝")
	fmt.Println()
	fmt.Println("This test verifies that document TTL expiration works correctly.")
	fmt.Println("We will insert documents with short TTL and verify they expire.")
	fmt.Println()

	// Load environment
	if err := godotenv.Load(); err != nil {
		log.Println("No .env file found, using defaults")
	}

	baseURL := os.Getenv("API_BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8080"
	}

	apiKey := os.Getenv("API_BASE_KEY")
	if apiKey == "" {
		apiKey = "a-test-api-key-from-ekodb"
	}

	// Create client
	client, err := ekodb.NewClient(baseURL, apiKey)
	if err != nil {
		return err
	}
	fmt.Println("✓ Client connected")

	collection := "ttl_expiration_test_go"
	ttlSeconds := 3
	defer func() {
		if err := client.DeleteCollection(collection); err != nil {
			runErr = errors.Join(runErr, fmt.Errorf("cleanup collection %s: %w", collection, err))
		} else {
			fmt.Println("✓ Deleted test collection")
		}
	}()

	// ═══════════════════════════════════════════════════════════════════════
	// TEST 1: Document TTL Expiration
	// ═══════════════════════════════════════════════════════════════════════
	fmt.Println()
	fmt.Println("═══════════════════════════════════════════════════════════")
	fmt.Println("TEST 1: Document TTL Expiration")
	fmt.Println("═══════════════════════════════════════════════════════════")

	// Step 1: Insert document with 3-second TTL
	fmt.Printf("\n[Step 1] Insert document with %d second TTL\n", ttlSeconds)
	fmt.Printf("  Input: {name: 'TTL Test', value: 'should expire'}\n")
	fmt.Printf("  TTL: %ds\n", ttlSeconds)

	record := ekodb.Record{
		"name":  "TTL Test",
		"value": "should expire",
	}
	doc, err := client.Insert(collection, record, ekodb.InsertOptions{
		TTL: fmt.Sprintf("%ds", ttlSeconds),
	})
	if err != nil {
		return fmt.Errorf("could not insert document: %w", err)
	}
	docID := doc["id"].(string)
	fmt.Printf("  Output: Document ID = %s\n", docID)
	fmt.Println("  ✓ PASS: Document inserted")

	// Step 2: Verify document exists immediately
	fmt.Println("\n[Step 2] Verify document exists immediately")
	fmt.Printf("  Input: FindByID(%s)\n", docID)

	found, err := client.FindByID(collection, docID)
	if err != nil {
		return fmt.Errorf("could not find document: %w", err)
	}
	if found == nil {
		return fmt.Errorf("document should exist but was not found")
	}
	fmt.Printf("  Output: Found document with name = %v\n", found["name"])
	fmt.Println("  ✓ PASS: Document exists")

	// Step 3: Wait for TTL to expire
	waitTime := time.Duration(ttlSeconds+2) * time.Second
	fmt.Printf("\n[Step 3] Wait for TTL to expire (%v)\n", waitTime)
	fmt.Printf("  Waiting")
	for i := 0; i < int(waitTime.Seconds()); i++ {
		time.Sleep(1 * time.Second)
		fmt.Printf(".")
	}
	fmt.Println(" done")
	fmt.Println("  ✓ PASS: Wait complete")

	// Step 4: Verify document is GONE
	fmt.Println("\n[Step 4] Verify document has expired")
	fmt.Printf("  Input: FindByID(%s)\n", docID)

	expired, err := client.FindByID(collection, docID)
	if err != nil {
		var httpErr *ekodb.HTTPError
		if !errors.As(err, &httpErr) || !httpErr.IsNotFound() {
			return fmt.Errorf("verify document expiration: %w", err)
		}
		fmt.Printf("  Output: Error (expected) - %v\n", err)
		fmt.Println("  ✓ PASS: Document expired (not found error)")
	} else if expired == nil {
		fmt.Println("  Output: nil (document not found)")
		fmt.Println("  ✓ PASS: Document expired correctly!")
	} else {
		fmt.Printf("  Output: Document still exists! %v\n", expired)
		return fmt.Errorf("document should have expired but still exists")
	}

	// ═══════════════════════════════════════════════════════════════════════
	// Cleanup
	// ═══════════════════════════════════════════════════════════════════════
	fmt.Println("\n═══════════════════════════════════════════════════════════")
	fmt.Println("CLEANUP")
	fmt.Println("═══════════════════════════════════════════════════════════")

	// ═══════════════════════════════════════════════════════════════════════
	// Summary
	// ═══════════════════════════════════════════════════════════════════════
	fmt.Println()
	fmt.Println("╔════════════════════════════════════════════════════════╗")
	fmt.Println("║              ALL TTL TESTS PASSED ✓                    ║")
	fmt.Println("╚════════════════════════════════════════════════════════╝")
	fmt.Println()
	fmt.Println("TTL expiration is working correctly:")
	fmt.Println("  • Documents with TTL expire after the specified time")
	fmt.Println("  • Documents without TTL persist indefinitely")
	fmt.Println("  • Different TTL durations are handled correctly")
	return nil
}

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}
